cns/restserver/ipam.go

```diff
@@ -92,26 +92,38 @@ func (service *HTTPRestService) releaseIPConfigHandler(w http.ResponseWriter, r
 	return
 }
 
-func (service *HTTPRestService) MarkIPsAsPending(numberToMark int) (map[string]cns.IPConfigurationStatus, error) {
+// MarkIPAsPendingRelease will mark IPs to pending release state.
+func (service *HTTPRestService) MarkIPAsPendingRelease(numberToMark int) (map[string]cns.IPConfigurationStatus, error) {
 	pendingReleaseIPs := make(map[string]cns.IPConfigurationStatus)
-	markedIPCount := 0
+	// Ensure PendingProgramming IPs will be release before Available ones.
+	ipStateTypes := [2]string{cns.PendingProgramming, cns.Available}
 
 	service.Lock()
 	defer service.Unlock()
-	for uuid, _ := range service.PodIPConfigState {
-		mutableIPConfig := service.PodIPConfigState[uuid]
-		if mutableIPConfig.State == cns.Available || mutableIPConfig.State == cns.PendingProgramming {
+	for _, ipStateType := range ipStateTypes {
+		pendingReleaseIPs = service.markSpecificIPTypeAsPending(numberToMark, len(pendingReleaseIPs), ipStateType)
+		alreadyMarkedNum := len(pendingReleaseIPs)
+		if alreadyMarkedNum == numberToMark {
+			return pendingReleaseIPs, nil
```

Comment:
I think this is still not correct. on the second iteration you are just returning the ips that went from available to pending. the previous ones got discarded. I would suggest the following structure for clarity

```go
allIPs := make(map[string]cns.IPConfigurationStatus)

for _, ipStateType := range ipStateTypes {
    releasedIPs := service.markSpecificIPTypeAsPending(numberToMark, ipStateType) // no need to send the second argument, just determine what the function needs from this scope
    /* 
        add releasedIPs to allIPs here
    */
    numberToMark -= len(releasedIPs) // determine the count for next iteration
    if numberToMark == 0 {
        return allIPs 
    }
}
```

Reply:
It works when we only have 2 ipStateTypes but not work when state type is over 2. Sure, I can change to calculate the number to mark independently.

Reply:
the problem is that the function call to `service.markSpecificIPTypeAsPending` returns a **new** map, so each assignment to `pendingReleaseIPs` overwrites the previous one. the second iteration will return the second allocated map.

Reply:
Yep I know but I think the current implementation already apvid it.